test/refract/test-JsonValue.cc

```diff
@@ -0,0 +1,1477 @@
+
+        WHEN("a JSON value is generated from it")
+        {
+            auto result = generateJsonValue(*el);
+
+            THEN("the result is `42`")
+            {
+                REQUIRE(to_string(result) == "42");
+            }
+        }
+    }
+
+    GIVEN("An empty NumberElement with sample and default")
+    {
+        auto el = make_empty<NumberElement>();
+        setDefault(*el, from_primitive(42));
+        addSample(*el, from_primitive(3));
+
+        WHEN("a JSON value is generated from it")
+        {
+            auto result = generateJsonValue(*el);
+
+            THEN("the result is `3`")
+            {
+                REQUIRE(to_string(result) == "3");
+            }
+        }
+    }
+
+    GIVEN("An empty, nullable NumberElement with sample and default")
+    {
+        auto el = make_empty<NumberElement>();
+        setDefault(*el, from_primitive(42));
+        addSample(*el, from_primitive(3));
+        setTypeAttribute(*el, "nullable");
+
+        WHEN("a JSON value is generated from it")
+        {
+            auto result = generateJsonValue(*el);
+
+            THEN("the result is `3`")
+            {
+                REQUIRE(to_string(result) == "3");
+            }
+        }
+    }
+
+    GIVEN("A nullable NumberElement with sample and default with content")
+    {
+        auto el = make_element<NumberElement>(5);
+        setDefault(*el, from_primitive(42));
+        addSample(*el, from_primitive(3));
+        setTypeAttribute(*el, "nullable");
+
+        WHEN("a JSON value is generated from it")
+        {
+            auto result = generateJsonValue(*el);
+
+            THEN("the result is `5`")
+            {
+                REQUIRE(to_string(result) == "5");
+            }
+        }
+    }
+
+    GIVEN("An empty NumberElement with a default")
+    {
+        auto el = make_empty<NumberElement>();
+        setDefault(*el, from_primitive(42));
+
+        WHEN("a JSON value is generated from it")
+        {
+            auto result = generateJsonValue(*el);
+
+            THEN("the result is `42`")
+            {
+                REQUIRE(to_string(result) == "42");
+            }
+        }
+    }
+
+    GIVEN("An empty, nullable NumberElement")
+    {
+        auto el = make_empty<NumberElement>();
+        setTypeAttribute(*el, "nullable");
+
+        WHEN("a JSON value is generated from it")
+        {
+            auto result = generateJsonValue(*el);
+
+            THEN("the result is `null`")
+            {
+                REQUIRE(to_string(result) == "null");
+            }
+        }
+    }
+
+    GIVEN("A NumberElement with content")
+    {
+        auto el = make_element<NumberElement>("42.1930e-39");
+
+        WHEN("a JSON value is generated from it")
+        {
+            auto result = generateJsonValue(*el);
+
+            THEN("the result is `42.1930e-39`")
+            {
+                REQUIRE(to_string(result) == "42.1930e-39");
+            }
+        }
+    }
+}
```

Comment:
Please add tests for `Nan` and `Inf`

Reply:
For NumberElement

Reply:
We use the JSON definition of number, `Nan` and `Inf` are not numbers.

Reply:
But you can set value of NumberElement to Nan/Inf, isn't it?
There should be, at least, defined bahavior for this case

Reply:
You can theoretically construct a NumberElement with any string now, as the check for JSON Number is in `Serialize.cc`. But it will have undefined behaviour, so I'm not sure what to test against. Finding something non-numberish in a NumberElement could be handled as an `assert` in `JsonValue.cc`.

Reply:
On other side refract is not strictly for JSON, so we should define what happens. e.g. conversion to zero?
Not sure, just idea

Reply:
There is the relevant issue on the MSON spec not being spec enough: https://github.com/apiaryio/mson/issues/93
In API Elements ("refract") we are pretty clear: https://github.com/apiaryio/api-elements/blob/master/docs/element-definitions.md#values
We could add a debug build run time check during JSON serialization (not JSON _value_ serialization), although I don't deem that necessary. Here are the tests currently covering your worries: https://github.com/apiaryio/drafter/blob/master/test/test-Serialize.cc#L931

Reply:
Ok. Let it pass for now and create issue to resolve this